### PR TITLE
MATLAB example

### PR DESCRIPTION
**Pull request by sbesson**
This PR builds on top the work started by @bramalingam to convert the MATLAB snippets included in the developer documentation (and tested as part of the release) into a fully executable standalone script. The synchronization between the code and the documentation is ensured using the same strategy as the one by @rleigh-dundee in the Bio-Formats C++ poject, i.e. examples are included under `docs/sphinx` until we find a better place and the reStructuredText makes use of `literalinclude` with delimiters in the Sphinx text to include snippets of code.

The CI job should be modified to test the `bftest.m` example file (better name?) on a daily basis and prevent docs for becoming out of sync. de44d68 and 2b2c169 also includes fixes to the various snippets reported by @bramalingam as part of 5.1.1 testing.

Additionally, MATLAB test are added for the colormap retrieval step in `bfopen` using fake files and the corresponding job should stay green.

### CODE REVIEW

File: docs/sphinx/developers/examples/bftest.m [hunk 1]
```diff
@@ -0,0 +1,169 @@
+bfCheckJavaPath();
+
+% Create local file for testing
+java_tmpdir = char(java.lang.System.getProperty('java.io.tmpdir'));
+uuid = char(java.util.UUID.randomUUID());
+tmpdir = fullfile(java_tmpdir, uuid);
+
```
Comment (manics) [inline docs/sphinx/developers/examples/bftest.m]: I needed to add `mkdir(tmpdir)` to get this running.

Comment (sbesson) [inline docs/sphinx/developers/examples/bftest.m]: Absolutely, this is likely the same issue as the one reported in https://ci.openmicroscopy.org/view/DEV/job/BIOFORMATS-DEV-merge-matlab/11/console.

**Comment by manics** [comment]
Comparing the text content of https://www.openmicroscopy.org/site/support/bio-formats5.1/developers/matlab-dev.html vs https://www.openmicroscopy.org/site/support/bio-formats5.1-staging/developers/matlab-dev.html:

```diff
--- bf.txt	2015-04-30 11:32:18.000000000 +0100
+++ bf-staging.txt	2015-04-30 11:32:18.000000000 +0100
@@ -55,7 +55,7 @@
       o
         Bio-Formats
 
-        </site/support/bio-formats5>
+        </site/support/bio-formats5.1-staging>
       o
         OME Model and Formats
 
@@ -162,7 +162,7 @@
   * index <../genindex.html>
   * next <python-dev.html> |
   * previous <conversion.html> |
-  * Bio-Formats 5.1.1 documentation <../index.html> »
+  * Bio-Formats 5.1.1-DEV documentation <../index.html> »
   * Developer Documentation <index.html> »
 
 
@@ -252,7 +252,7 @@
     Opening an image file¶ <#opening-an-image-file>
 
 The first thing to do is initialize a file with the bfopen
-<https://github.com/openmicroscopy/bioformats/blob/v5.1.1/components/formats-gpl/matlab/bfopen.m>
+<https://github.com/snoopycrimecop/bioformats/blob/merge/develop/latest/components/formats-gpl/matlab/bfopen.m>
 function:
 
 data = bfopen('/path/to/data/file');
@@ -281,13 +281,12 @@
 
 Here is an example of how to unwrap specific image planes for easy access:
 
-data = bfopen('/path/to/data/file');
 seriesCount = size(data, 1);
 series1 = data{1, 1};
 series2 = data{2, 1};
 series3 = data{3, 1};
 metadataList = data{1, 2};
-% …etc.
+% etc
 series1_planeCount = size(series1, 1);
 series1_plane1 = series1{1, 1};
 series1_label1 = series1{1, 2};
@@ -295,24 +294,18 @@
 series1_label2 = series1{2, 2};
 series1_plane3 = series1{3, 1};
 series1_label3 = series1{3, 2};
-% …etc.
 
 
       Displaying images¶ <#displaying-images>
 
 If you want to display one of the images, you can do so as follows:
 
-data = bfopen('/path/to/data/file');
-% plot the 1st series's 1st image plane in a new figure
-series1 = data{1, 1};
-series1_plane1 = series1{1, 1};
-series1_label1 = series1{1, 2};
 series1_colorMaps = data{1, 3};
 figure('Name', series1_label1);
 if (isempty(series1_colorMaps{1}))
   colormap(gray);
 else
-  colormap(series1_colorMaps{1});
+  colormap(series1_colorMaps{1}(1,:));
 end
 imagesc(series1_plane1);
 
@@ -355,7 +348,6 @@
 
 To retrieve the metadata value for specific keys:
 
-data = bfopen('/path/to/data/file');
 % Query some metadata fields (keys are format-dependent)
 metadata = data{1, 2};
 subject = metadata.get('Subject');
@@ -363,8 +355,6 @@
 
 To print out all of the metadata key/value pairs for the first series:
 
-data = bfopen('/path/to/data/file');
-metadata = data{1, 2};
 metadataKeys = metadata.keySet().iterator();
 for i=1:metadata.size()
   key = metadataKeys.nextElement();
@@ -381,7 +371,6 @@
 
 To access physical voxel and stack sizes of the data:
 
-data = bfopen('/path/to/data/file');
 omeMeta = data{1, 4};
 stackSizeX = omeMeta.getPixelsSizeX(0).getValue(); % image width, pixels
 stackSizeY = omeMeta.getPixelsSizeY(0).getValue(); % image height, pixels
@@ -390,11 +379,8 @@
 voxelSizeXdefaultValue = omeMeta.getPixelsPhysicalSizeX(0).value();           % returns value in default unit
 voxelSizeXdefaultUnit = omeMeta.getPixelsPhysicalSizeX(0).unit().getSymbol(); % returns the default unit type
 voxelSizeX = omeMeta.getPixelsPhysicalSizeX(0).value(ome.units.UNITS.MICROM); % in µm
-voxelSizeXdouble = voxelSizeX.doubleValue();                                  % The numeric value represented by this object after conversion to type double
 voxelSizeY = omeMeta.getPixelsPhysicalSizeY(0).value(ome.units.UNITS.MICROM); % in µm
-voxelSizeYdouble = voxelSizeY.doubleValue();                                  % The numeric value represented by this object after conversion to type double
 voxelSizeZ = omeMeta.getPixelsPhysicalSizeZ(0).value(ome.units.UNITS.MICROM); % in µm
-voxelSizeZdouble = voxelSizeZ.doubleValue();                                  % The numeric value represented by this object after conversion to type double
 
 For more information about the methods to retrieve the metadata, see the
 MetadataRetrieve
@@ -409,13 +395,13 @@
     Reading from an image file¶ <#reading-from-an-image-file>
 
 The main inconvenience of the bfopen.m
-<https://github.com/openmicroscopy/bioformats/blob/v5.1.1/components/formats-gpl/matlab/bfopen.m>
+<https://github.com/snoopycrimecop/bioformats/blob/merge/develop/latest/components/formats-gpl/matlab/bfopen.m>
 function is that it loads all the content of an image regardless of its
 size.
 
 To access the file reader without loading all the data, use the
 low-level bfGetReader.m
-<https://github.com/openmicroscopy/bioformats/blob/v5.1.1/components/formats-gpl/matlab/bfGetReader.m>
+<https://github.com/snoopycrimecop/bioformats/blob/merge/develop/latest/components/formats-gpl/matlab/bfGetReader.m>
 function:
 
 reader = bfGetReader('path/to/data/file');
@@ -425,7 +411,7 @@
 omeMeta = reader.getMetadataStore();
 
 Individual planes can be queried using the bfGetPlane.m
-<https://github.com/openmicroscopy/bioformats/blob/v5.1.1/components/formats-gpl/matlab/bfGetPlane.m>
+<https://github.com/snoopycrimecop/bioformats/blob/merge/develop/latest/components/formats-gpl/matlab/bfGetPlane.m>
 function:
 
 series1_plane1 = bfGetPlane(reader, 1);
@@ -448,38 +434,37 @@
 
 The basic code for saving a 5D array into an OME-TIFF file is located in
 the bfsave.m
-<https://github.com/openmicroscopy/bioformats/blob/v5.1.1/components/formats-gpl/matlab/bfsave.m>
+<https://github.com/snoopycrimecop/bioformats/blob/merge/develop/latest/components/formats-gpl/matlab/bfsave.m>
 function.
 
 For instance, the following code will save a single image of 64 pixels
 by 64 pixels with 8 unsigned bits per pixels:
 
 plane = zeros(64, 64, 'uint8');
-bfsave(plane, 'my-file.ome.tiff');
+bfsave(plane, 'single-plane.ome.tiff');
 
 And the following code snippet will produce an image of 64 pixels by 64
 pixels with 2 channels and 2 timepoints:
 
 plane = zeros(64, 64, 1, 2, 2, 'uint8');
-bfsave(plane, 'my-file.ome.tiff');
+bfsave(plane, 'multiple-planes.ome.tiff');
 
 By default, bfsave will create a minimal OME-XML metadata object
 containing basic information such as the pixel dimensions, the dimension
 order and the pixel type. To customize the OME metadata, it is possible
 to create a metadata object from the input array using
 createMinimalOMEXMLMetadata.m
-<https://github.com/openmicroscopy/bioformats/blob/v5.1.1/components/formats-gpl/matlab/createMinimalOMEXMLMetadata.m>,
+<https://github.com/snoopycrimecop/bioformats/blob/merge/develop/latest/components/formats-gpl/matlab/createMinimalOMEXMLMetadata.m>,
 add custom metadata and pass this object directly to bfsave:
 
 plane = zeros(64, 64, 1, 2, 2, 'uint8');
 metadata = createMinimalOMEXMLMetadata(plane);
-pixelSize = ome.xml.model.primitives.PositiveFloat(java.lang.Double(.05));
+pixelSize = ome.units.quantity.Length(java.lang.Double(.05), ome.units.UNITS.MICROM);
 metadata.setPixelsPhysicalSizeX(pixelSize, 0);
 metadata.setPixelsPhysicalSizeY(pixelSize, 0);
-pixelSizeZ = ome.xml.model.primitives.PositiveFloat(java.lang.Double(.2));
+pixelSizeZ = ome.units.quantity.Length(java.lang.Double(.2), ome.units.UNITS.MICROM);
 metadata.setPixelsPhysicalSizeZ(pixelSizeZ, 0);
-...
-bfsave(plane, 'my-file.ome.tiff', 'metadata', metadata);
+bfsave(plane, 'metadata.ome.tiff', 'metadata', metadata);
 
 For more information about the methods to store the metadata, see the
 MetadataStore
@@ -510,12 +495,12 @@
 % Construct an empty Bio-Formats reader
 r = bfGetReader();
 % Decorate the reader with the Memoizer wrapper
-r = loci.formats.Memoizer(r)
+r = loci.formats.Memoizer(r);
 % Initialize the reader with an input file
 % If the call is longer than a minimal time, the initialized reader will
 % be cached in a file under the same directory as the initial file
 % name .large_file.bfmemo
-r.setId('/path/to/large_file');
+r.setId(pathToFile);
 
 % Perform work using the reader
 
@@ -525,7 +510,7 @@
 % If the reader has been cached in the call above, re-initializing the
 % reader will use the memo file and complete much faster especially for
 % large data
-r.setId('/path/to/large_file');
+r.setId(pathToFile);
 
 % Perform additional work
 
@@ -554,19 +539,21 @@
 MATLAB parfor to do work on a single file in a parallel loop:
 
 % Construct a Bio-Formats reader decorated with the Memoizer wrapper
-r = loci.formats.Memoizer(bfGetReader(r, 0))
+r = loci.formats.Memoizer(bfGetReader(), 0);
 % Initialize the reader with an input file to cache the reader
-r.setId('/path/to/large_file');
+r.setId(pathToFile);
 % Close reader
 r.close()
 
+nWorkers = 4;
+
 % Enter parallel loop
-parfor i = 1 : nWorkers
+for i = 1 : nWorkers
     % Initialize a new reader per worker as Bio-Formats is not thread safe
-    r2 = loci.formats.Memoizer(bfGetReader(r, 0))
+    r2 = loci.formats.Memoizer(bfGetReader(), 0);
     % Initialization should use the memo file cached before entering the
     % parallel loop
-    r2.setId('/path/to/large_file');
+    r2.setId(pathToFile);
 
     % Perform work
 
@@ -580,8 +567,8 @@
   * index <../genindex.html>
   * next <python-dev.html> |
   * previous <conversion.html> |
-  * Bio-Formats 5.1.1 documentation <../index.html> »
+  * Bio-Formats 5.1.1-DEV documentation <../index.html> »
   * Developer Documentation <index.html> »
 
 © Copyright 2000-2015, The Open Microscopy Environment . Last generated
-on Apr 28, 2015. Created using Sphinx <http://sphinx.pocoo.org/> 1.2.3.
+on Apr 30, 2015. Created using Sphinx <http://sphinx.pocoo.org/> 1.2.3.
```

**Comment by sbesson** [comment]
Last commits should fix the status of https://ci.openmicroscopy.org/job/BIOFORMATS-5.1-merge-matlab/

**Comment by jburel** [comment]
job is green again

**Comment by sbesson** [comment]
@jburel: pushed cleanup commit. Can you relist for re-review or clear it yourself tomorrow?

**Comment by jburel** [comment]
@sbesson: done

**Comment by jburel** [comment]
Thanks for changes, looking good

**Pull request merged**